### ISSUE TITLE
Feature request: option to format on save

### ISSUE THREAD

**Issue by chenglou** [facebook/nuclide#504]
Nuclide's format-js has that. However, I see it's marked as "not recommended yet" for some reason. Is there a rationale behind it?

**Comment by shushz** [comment]
I assume that the rationale is this functionality being not mature enough yet to recommend its wide usage.
/cc @kyldvs

**Comment by chenglou** [comment]
Btw, I tried implementing this using the existing nuclide code. I can't intercept the editor on `onWillSave` and paste in the newly formatted code; setting the buffer to the formatted code seems to happen after the save. I'm guessing it's because API for formatting returns a promise and I resolve that and do the formatting, but the save's already done? The format-js code doesn't use the nuclide formatting facility afaik, so didn't have this problem.

**Comment by kyldvs** [comment]
Yeah I'm hesitant to add this functionality because it's definitely not mature enough. I still expect that there will be issues every so often and the user should manually review which requires get added. Making it run on every save seems dangerous to me.

(also on a more practical note it's not integrated with grammars, it can be run on any file type, it just happens to throw a parse error on non-js files. That functionality would need to be built so that it only runs-on-save for JS files too)

There is save-after-run functionality built into it though, which is kind of similar.

**Comment by nmote** [comment]
We have the `nuclide-code-format` package which is language-agnostic. For some reason it doesn't appear that format-js is integrated with it. But anyway, that would be the right place to add format-on-save if we wanted it. Generally-speaking we want features like this to be language-agnostic and to have separate packages that plug in for each language.

I'm not sure what the implementation would look like, though. We wouldn't want to block the file save on a potentially expensive and error-prone operation. Maybe we would save again after the results come back? However that comes with its own set of problems.

**Comment by tyangliu** [comment]
@nmote Oh this doesn't relate to the format-js case, but just for tracking - I implemented format-on-save in nuclide-code-format [here](https://github.com/facebook/nuclide/commit/e34527cecb478535624d540d5efe058666982e87). It listens to onDidSave, formats, then resaves after formatting, but only does so if the file contents haven't changed between before and after the formatting call.
